Add new SYCL language spec for fpga_mem (#9533)
intel/llvm: sycl/doc/extensions/proposed/sycl_ext_intel_fpga_datapath.asciidoc

// for syntax highlighting purposes. This is needed because
// docbook uses c++ and html5 uses cpp.
:language: {basebackend@docbook:c++:cpp}

== Notice

[%hardbreaks]
Copyright (C) 2023-2023 Intel Corporation. All rights reserved.

Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
of The Khronos Group Inc. OpenCL(TM) is a trademark of Apple Inc. used by
permission by Khronos.

== Contact

To report problems with this extension, please open a new issue at:

https://github.com/intel/llvm/issues

== Dependencies

This extension is written against the SYCL 2020 revision 7 specification. All
references below to the "core SYCL specification" or to section numbers in the
SYCL specification refer to that revision.

This extension also depends on the following other SYCL extensions:

* link:../supported/sycl_ext_oneapi_accessor_properties.asciidoc[
  sycl_ext_oneapi_accessor_properties.asciidoc] as described in the <<Properties for fpga memory>> section below
* link:sycl_ext_intel_fpga_mem.asciidoc[sycl_ext_intel_fpga_mem.asciidoc] as described in the <<Restrictions on creating fpga_mem objects>> section below

> * link:sycl_ext_oneapi_device_global.asciidoc[sycl_ext_oneapi_device_global.asciidoc] as described in the <<Restrictions on creating fpga_mem objects>> section below

Reviewer: This extension doesn't depend on the device_global extension, it merely has an interaction with it that is spelled out below.  Can you make that clear here?

== Status
Experimental

This is an experimental extension specification, intended to provide early
access to features and gather community feedback. Interfaces defined in this

> specification are implemented in {dpcpp}, but they are not finalized and may

Reviewer: They are not yet implemented because this is only a proposal.

member of `fpga_datapath` allows a reference to be extracted, to which the usual
dot operator may be applied.

== Specification

`fpga_datapath` is a class template templated on type T, that represents an object of type T. It is a request to the compiler to implement that object, when present in FPGA device code, in the datapath, as opposed to an off-datapath memory.

=== Feature test macro

This extension provides a feature-test macro as described in the core SYCL
specification. An implementation supporting this extension must predefine the
macro `SYCL_EXT_INTEL_FPGA_DATAPATH` to one of the values defined in the table
below. Applications can test for the existence of this macro to determine if
the implementation supports this feature, or applications can test the macro's
value to determine which of the extension's features the implementation
supports.

[%header,cols="1,5"]
|===
|Value
|Description

|1
|The APIs of this experimental extension are not versioned, so the
 feature-test macro always has this value.
|===

=== Representation of FPGA memory

[source,c++]
----
namespace sycl::ext::intel::experimental {

> template <typename T, typename PropListT = sycl::ext::oneapi::properties<>>

Author: typo. This is not mentioned in the description or in the header code sample.

    // "T" need only define these operators as non-member functions. Because
    // there is an implicit conversion from "fpga_datapath" to "T&", the operations
    // can be applied to objects of type "fpga_datapath<T>"
  };

} // namespace sycl::ext::intel::experimental
----

[frame="topbot",options="header"]
|===
|Functions |Description

// --- ROW BREAK ---
a|
[source,c++]
----
template<typename ... S>
constexpr fpga_datapath(S ... args);
----
|
Constructs a `fpga_datapath` object, and implicit storage for `T`.

// --- ROW BREAK ---
a|
[source,c++]
----
operator T&() noexcept;
operator const T&() const noexcept;
----
|
Implicit conversion to a reference to the underlying `T`.

// --- ROW BREAK ---
a|
[source,c++]
----
T& get() noexcept;
const T& get() const noexcept;
----
|
Returns a reference to the underlying `T`.

|===

=== Restrictions on creating fpga_datapath objects

There are restrictions on how the application can create objects of type
`fpga_datapath`. Applications that violate these restrictions are ill-formed.

* The `fpga_datapath` variable can only have accesses from a single kernel. This happens when `device_global` is templated on a `fpga_datapath` type. 

> * The `fpga_datapath` variable cannot be templated on a `fpga_mem` object, as this is a conflict of requirements on how the variable should be implemented, on or off the datapath. Diagnostic is required.

Reviewer: This simplistic check doesn't catch all the cases that don't make sense because one could easily have the fpga_mem wrapped in another object.  For example:

```
class foo {
  fpga_mem<int> a;
  int b;
};
...
q.single_task([=] {
  fpga_datapath<foo> c;
});
```

But addressing this comprehensively is complicated as you have to recursively inspect all the member of any type that an fpga_datapath is templated on.  I don't even think that could be done without compiler support which would preclude a purely C++-based implementation of the extension.  It would have to be implemented in either _every_ backend, even non-FPGA ones, which is inefficient, or in the frontend.

I don't think this effort is worth it.  Could we instead simply create a simple rule for how the two extensions interact?  For example, we could state that the more narrow annotation wins.  In my example above, that would mean that a would be implemented in memory while b would be implemented in the datapath.

Author: I have defined the behaviour of templating `fpga_mem` on `fpga_datapath` or vice versa. 

As for having arrays of `fpga_mem` i eventually concluded that it should be allowed. Since it seems arbitrary to prevent aggregates of a class. I don't think any special notes about it should be included in the SYCL spec. We can call out this behavior in the FPGA specific documentation.

= sycl_ext_intel_fpga_datapath

:source-highlighter: coderay
:coderay-linenums-mode: table

// This section needs to be after the document title.
:doctype: book
:toc2:
:toc: left
:encoding: utf-8
:lang: en
:dpcpp: pass:[DPC++]

// Set the default source code type in this document to C++,
// for syntax highlighting purposes. This is needed because
// docbook uses c++ and html5 uses cpp.
:language: {basebackend@docbook:c++:cpp}

== Notice

[%hardbreaks]
Copyright (C) 2023-2023 Intel Corporation. All rights reserved.

Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
of The Khronos Group Inc. OpenCL(TM) is a trademark of Apple Inc. used by
permission by Khronos.

== Contact

To report problems with this extension, please open a new issue at:

https://github.com/intel/llvm/issues

== Dependencies

This extension is written against the SYCL 2020 revision 7 specification. All
references below to the "core SYCL specification" or to section numbers in the
SYCL specification refer to that revision.

> This extension does not depend on other SYCL extensions, it is self-contained. Below is a list of interactions with other SYCL extensions. These extensions don't have to be implemented to support the `sycl_ext_intel_fpga_mem` extension.

Reviewer: ```suggestion
This extension does not depend on other SYCL extensions, it is self-contained. Below is a list of interactions with other SYCL extensions. These extensions don't have to be implemented to support the `sycl_ext_intel_fpga_datapath` extension.
```

> This extension does not depend on other SYCL extensions, it is self-contained. Below is a list of interactions with other SYCL extensions. These extensions don't have to be implemented to support the `sycl_ext_intel_fpga_datapath` extension.

Reviewer: ```suggestion
This extension does not depend on other SYCL extensions; it is self-contained. Below is a list of interactions with other SYCL extensions. These extensions don't have to be implemented to support the `sycl_ext_intel_fpga_datapath` extension.
```

// This section needs to be after the document title.
:doctype: book
:toc2:
:toc: left
:encoding: utf-8
:lang: en
:dpcpp: pass:[DPC++]

// Set the default source code type in this document to C++,
// for syntax highlighting purposes. This is needed because
// docbook uses c++ and html5 uses cpp.
:language: {basebackend@docbook:c++:cpp}

== Notice

[%hardbreaks]
Copyright (C) 2023-2023 Intel Corporation. All rights reserved.

Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
of The Khronos Group Inc. OpenCL(TM) is a trademark of Apple Inc. used by
permission by Khronos.

== Contact

To report problems with this extension, please open a new issue at:

https://github.com/intel/llvm/issues

== Dependencies

This extension is written against the SYCL 2020 revision 7 specification. All
references below to the "core SYCL specification" or to section numbers in the
SYCL specification refer to that revision.

This extension does not depend on other SYCL extensions, it is self-contained. Below is a list of interactions with other SYCL extensions. These extensions don't hve to be implemented to support the `sycl_ext_intel_fpga_mem` extension.

> * link:sycl_ext_intel_fpga_mem.asciidoc[sycl_ext_intel_fpga_mem.asciidoc] as described in the <<Restrictions on creating fpga_mem objects>> section below
> * link:sycl_ext_oneapi_device_global.asciidoc[sycl_ext_oneapi_device_global.asciidoc] as described in the <<Restrictions on creating fpga_mem objects>> section below

Reviewer: Links are wrong:
```suggestion
* link:sycl_ext_intel_fpga_mem.asciidoc[sycl_ext_intel_fpga_mem.asciidoc] as described in the <<Restrictions on creating fpga_datapath objects>> section below
* link:sycl_ext_oneapi_device_global.asciidoc[sycl_ext_oneapi_device_global.asciidoc] as described in the <<Restrictions on creating fpga_datapath objects>> section below
```

== Status
Proposed

This is a proposed extension specification, intended to gather community
feedback.  Interfaces defined in this specification may not be implemented yet
or may be in a preliminary state.  The specification itself may also change in
incompatible ways before it is finalized. *Shipping software products should
not rely on APIs defined in this specification.*

== Overview

[NOTE]
====
In this document, we use `fpga_datapath` to indicate the proposed `sycl::ext::intel::experimental::fpga_datapath`.
====

The purpose of this document is to clearly describe and specify `fpga_datapath` and 
related concepts, types, and mechanisms, and to give examples and context for their usage.

=== Motivation
`fpga_datapath` is a class template templated on a type T, that represents an object of type T. It is a request to the compiler to implement that object, when present in FPGA device code, in the datapath, as opposed to an off-datapath memory.

=== Example
[source,c++]
----
struct MyClass {
  bool x;
};

using namespace sycl::ext::intel::experimental;

sycl::queue q;
q.single_task([=] {

>   fpga_datapath<int[4]> fm1{1, 3, 5, 7};
>   fpga_datapath<MyClass> fm2;

Reviewer: Not strictly necessary, but it makes more sense:
```suggestion
  fpga_datapath<int[4]> fd1{1, 3, 5, 7};
  fpga_datapath<MyClass> fd2;
```

  fm2.get().x = fm1[0];
});
----

> Above we see an example that declares `fpga_datapath` variables `fm1`, `fm2`, and accesses them. 
> `fm1` creates an allocation on the device that is private to the kernel and contains an array of four `int` types that is initialized to `{1, 3, 5, 7}`. `fm2` creates an allocation on the device that is private to the kernel and contains an object of type `MyClass` that is default-initialized.

Reviewer: fm->fd
also "four 'int' types" sounds off.  That makes it sound like an array of types rather than an array of objects.  
```suggestion
Above we see an example that declares `fpga_datapath` variables `fd1`, `fd2`, and accesses them. 
`fd1` creates an allocation on the device that is private to the kernel and contains an array of four `int`s that is initialized to `{1, 3, 5, 7}`. `fd2` creates an allocation on the device that is private to the kernel and contains an object of type `MyClass` that is default-initialized.
```

because {cpp} doesn't allow for overloading of the "dot operator", a `get()`
member of `fpga_datapath` allows a reference to be extracted, to which the usual
dot operator may be applied.

== Specification

`fpga_datapath` is a class template templated on a type T, that represents an object of type T. It is a request to the compiler to implement that object, when present in FPGA device code, in the datapath, as opposed to an off-datapath memory.

=== Feature test macro

This extension provides a feature-test macro as described in the core SYCL
specification. An implementation supporting this extension must predefine the
macro `SYCL_EXT_INTEL_FPGA_DATAPATH` to one of the values defined in the table
below. Applications can test for the existence of this macro to determine if
the implementation supports this feature, or applications can test the macro's
value to determine which of the extension's features the implementation
supports.

[%header,cols="1,5"]
|===
|Value
|Description

|1
|The APIs of this experimental extension are not versioned, so the
 feature-test macro always has this value.
|===

> === Representation of FPGA memory

Reviewer: ```suggestion
=== Representation of FPGA datapath
```

The section below and the table following describe the constructors and member functions for `fpga_datapath`.

[source,c++]
----
namespace sycl::ext::intel::experimental {

  template <typename T>
  class fpga_datapath {
  public:
    template <typename... S> constexpr fpga_datapath(S... args);
    
    // Access the underlying data
    operator T &() noexcept;
    operator const T &() const noexcept;

    T &get() noexcept;
    const T &get() const noexcept;

    // Enable assignments from underlying type
    fpga_datapath &operator=(const T &) noexcept;

    // Note that there is no need for "fpga_datapath" to define member functions
    // for operators like "++", "[]", "->", comparison, etc. Instead, the type
    // "T" need only define these operators as non-member functions. Because
    // there is an implicit conversion from "fpga_datapath" to "T&", the operations
    // can be applied to objects of type "fpga_datapath<T>"
  };

} // namespace sycl::ext::intel::experimental
----

[frame="topbot",options="header"]
|===
|Functions |Description

// --- ROW BREAK ---
a|
[source,c++]
----
template<typename ... S>
constexpr fpga_datapath(S ... args);
----
|

> Constructs a `fpga_datapath` object, and implicit storage for `T`.

Reviewer: It's worth mentioning that this calls T's constructor with arguments "args..." to initialize the underlying data.

// --- ROW BREAK ---
a|
[source,c++]
----
operator T&() noexcept;
operator const T&() const noexcept;
----
|
Implicit conversion to a reference to the underlying `T`.

// --- ROW BREAK ---
a|
[source,c++]
----
T& get() noexcept;
const T& get() const noexcept;
----
|
Returns a reference to the underlying `T`.

|===

=== Restrictions on creating fpga_datapath objects

There are restrictions on how the application can create objects of type
`fpga_datapath`. Applications that violate these restrictions are ill-formed.

> * The `fpga_datapath` variable can only have accesses from a single kernel. This can happen when `device_global` is templated on a `fpga_datapath` type. 

Reviewer: I don't think we need this restriction in the spec.  We've already said that the class is a "request", implying that an implementation can ignore it in some cases.  We could document in our docs that this is one case in which our implementation ignores it.  I worry that with this wording ("... are ill-formed"), the frontend folks will interpret it as their responsibility to error out on any program that accesses such a variable in multiple kernels which is not typically a task you'd want to perform in the FE.

Reviewer: Note that if you remove this, you should also remove the reference to the device_global extension in the dependencies section because this is the only mention of device_global in the spec.

Author: I decided to call it out explicitly. I think that is clearer for understanding.

Reviewer: After thinking about it more, I don't think this is implementable as written.  If the code is ill-formed (without the "no diagnostic required" qualifier) then the FE must diagnose it.  But this situation is impossible for the FE to diagnose without building a call graph of the kernel (which the FE team has already told us they don't think should be the responsibility of the FE) and is completely impossible to do at compile time at all for any target that supports the function pointers extension.  

Furthermore, we don't need to say this in the spec and it adds significant complexity to the spec by creating a link to device_globals.  We've already said that the class is a hint and so we are within our rights to not follow it in some cases.  We should instead mention this case in our docs and emit a warning when the user hits it.  This also future-proofs the spec against possible improvements to the compiler that would allow fpga_datapath variables to be accessed by more than one kernel (for example if we supported kernel fusion).  In such a case we'd merely have to update or docs and implementation rather than the spec itself.

Author: removed.
